File: qml-intro/ex-simple-item/pyside/ellipseitem.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
# Copyright (c) 2019, Klaralvdalens Datakonsult AB (KDAB)
# All rights reserved.
#
# See the LICENSE.txt file shipped along with this file for the license.

from PySide2.QtCore import Signal, Property
from PySide2.QtGui import QPen, QColor
from PySide2.QtQuick import QQuickPaintedItem


class EllipseItem(QQuickPaintedItem):
    colorChanged = Signal()

    def __init__(self, parent = None):
        QQuickPaintedItem.__init__(self, parent)
        self._color = None

    def paint(self, painter):
        painter.save()
        pen = QPen(self._color)
        pen.setWidth(3)
        painter.setPen(pen)
        painter.drawEllipse(0, 0, self.width(), self.height())
        painter.restore()

    def getColor(self):
        return self._color

    def setColor(self, new_color):
        if self._color != new_color:
            self._color = new_color
            self.update()
            self.colorChanged.emit()

    color = Property(QColor, getColor, setColor, notify=colorChanged)

File: qml-intro/ex-simple-item/cpp/ellipseitem.cpp
/*************************************************************************
 *
 * Copyright (c) 2008-2011, Nokia Corporation and/or its subsidiary(-ies).
 * Copyright (c) 2019, Klaralvdalens Datakonsult AB (KDAB)
 * All rights reserved.
 *
 * See the LICENSE.txt file shipped along with this file for the license.
 *
 *************************************************************************/

#include <QPainter>
#include "ellipseitem.h"

//--> slide
EllipseItem::EllipseItem(QQuickItem *parent)
    : QQuickPaintedItem(parent)
{
}

void EllipseItem::paint(QPainter *painter)
{
//--> hide
    painter->save();

    QPen pen = painter->pen();
    pen.setBrush(m_color);
    pen.setWidth(3);
    painter->setPen(pen);
    const QRectF paintRect = QRectF({}, QSizeF{width(), height()}).adjusted(1, 1, -1, -1);
//<-- hide
    painter->drawEllipse(paintRect);
//--> hide
    painter->restore();
//<-- hide
}
//<-- slide

QColor EllipseItem::color() const
{
    return m_color;
}

void EllipseItem::setColor(const QColor &newColor)
{
    if (m_color != newColor) {
        m_color = newColor;
        update();
        emit colorChanged();
    }
}
